### PR TITLE
Fuzzy search using Levenshtein distance

### PR DESCRIPTION
**Pull request by arturz**

### CODE REVIEW

File: lib/kanta/migrations/postgresql/v01.ex
Comment (ravensiris) [inline lib/kanta/migrations/postgresql/v01.ex:1]: This breaks the versioning system.
If I already migrated the v1 and ran this again I would be left off with the old migration.
Instead regenerate this field in v2 migration.

Also is it really `'english'`? Default lang could be something other than english(not sure if kanta expects only english).
It would be nice to search with the msg_id, po translation and kanta translation with highest weight(A) on msg_id, then lower on kanta translation then on po translation. Also you might want to include the domain as the lowest ranking.

File: lib/kanta/migrations/postgresql/v02.ex [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    execute("CREATE EXTENSION IF NOT EXISTS pg_trgm;")
+    execute("CREATE EXTENSION IF NOT EXISTS fuzzystrmatch;")
+
+    execute """
+      CREATE INDEX IF NOT EXISTS #{@kanta_messages}_msgid_gin_idx ON #{prefix}.#{@kanta_messages} USING gin(msgid gin_trgm_ops);
+    """
+
+    execute """
+      CREATE INDEX IF NOT EXISTS #{@kanta_messages}_msgid_soundex_idx ON #{prefix}.#{@kanta_messages} (soundex(msgid));
+    """
```
Comment (ravensiris) [inline lib/kanta/migrations/postgresql/v02.ex:40]: if we have more than 1 schema would this still work? no name collision for the index name?
maybe omit the name and leave it to postgres? https://www.postgresql.org/docs/current/sql-createindex.html says that a sensible default name would be chosen when not provided.

File: lib/kanta/query.ex [hunk 2]
```diff
@@ -293,6 +294,31 @@ defmodule Kanta.Query do
...
+      end
+
+      defp search_query_fuzzy(query, search) do
+        from(s in unquote(opts[:module]),
+          where:
+            fragment(
+              "msgid ILIKE ? OR (SOUNDEX(msgid) = SOUNDEX(?) AND LEVENSHTEIN(?, msgid) <= 3)",
+              ^("%" <> search <> "%"),
+              ^search,
+              ^search
```
Comment (ravensiris) [inline lib/kanta/query.ex:313]: I don't see `searchable` mentioned here?
You should form the query `where` using `to_tsquery()` and use `@@` operator for matching against the `searchable`.
Then order the results using `ts_rank(...)`

Example:

```elixir
  defmacro form_search_query(search_term) do
    quote do
      fragment(
        "SELECT to_tsquery(string_agg(unaccent(lexeme) || ':*', ' & ' order by positions)) FROM unnest(to_tsvector(?))",
        unquote(search_term)
      )
    end
  end

  defmacro ts_rank(left, right) do
    quote do
      fragment("ts_rank(?, ?)", unquote(left), unquote(right))
    end
  end

# then somewhere
query
|> 
where(
          [thing_that_holds_searchable: thing],
          fragment("? @@ ?", thing.searchable, form_search_query(^search_term))
        )
|> order_by(
          [thing_that_holds_searchable: thing],
          desc:
            ts_rank(
              thing.searchable,
              form_search_query(^search_term)
            )
        )
```

**Review by ravensiris — COMMENTED** [review]
I think this would need a tests with some edge cases.
E.g.
What happens when you search with spaces between?
"err mes" -> would it find a message like "error_message" ? You might need to tweak for spaces.
"hęłłó" -> would it find a message like "hello" ? Unaccent might be needed for this.
"wrd" -> would it find message like "word" or "world"?

I think adding a search by kanta/po translations would be very useful.
Also make sure that `searchable` gets refreshed properly when changes are made to translations.
